### ISSUE TITLE
[FEA] provide external libraries a way of getting a `DeviceBuffer` pointer that can become spillable again

### ISSUE THREAD

**Issue by wence-** [rapidsai/cudf#14029]
**Is your feature request related to a problem? Please describe.**

When running in a multi-gpu setting, message passing with ucx-py takes a `DeviceBuffer` and obtains the device memory pointer through the `__cuda_array_interface__`. This, correctly, marks the buffer as unspillable. 

It would be nice if there were a way to expose a pointer that is marked as unspillable until the external library drops the reference (kind of like `acquire_spill_lock`). `ucx-py` could then use it, and scope the pointer use to the lifetime of the message request (once the request is completed, the pointer can be dropped and is available for spilling again).

**Describe the solution you'd like**

If we were to hand back an object that had a `weakref.finalize(obj, unmark_spillable)` callback, when it was dropped, we could let the buffer be spillable again.

**Describe alternatives you've considered**

Making ucx-py aware of cudf and using `acquire_spill_lock`.

cc @madsbk / @vyasr / @galipremsagar

**Comment by madsbk** [comment]
Agree, this would be very useful but notice we do something like this already when serializing. 
`SpillableBuffer.serialize()` returns a `Buffer` with the *spill lock* as owner:
https://github.com/rapidsai/cudf/blob/branch-23.10/python/cudf/cudf/core/buffer/spillable_buffer.py#L460-L474

**Comment by wence-** [comment]
Ah, I think I hadn't spotted that. That also works I think, so things are less bad than I thought

**Comment by vyasr** [comment]
I think that we need to be a bit more careful regarding types. We cannot assume that the underlying data source is a DeviceBuffer since it could be externally owned memory (and after the recent changes to arrow interop, our special arrow owner types), so we need to ensure that we keep our abstractions from leaking and only hand out information at the level of a gpumemoryview or higher without reaching into it to grab the underlying data type.
